Structured audit logging

albertzaharovits opened the pull request:
As it now stands logfile security auditing is *almost* structured. Most fields are of the form `key=[value]` and there is no text in between the entries (besides the comma and space delimiters). If all values where of the form `key=[value]` then such entries would count as `'structured'`. Examples of such values, which are *not* prepended by the field name, are: `date`, `host name`, `node name`, `layer: rest or transport`.

This PR builds the log event as a *Map*, and lets log4j's internals (precisely the `layout` assigned to the `appender`) to handle the actual printing format. This is better because we could have many appenders for the same log each with a particular layout from: JSON, XML or syslog (all structured) to custom ones to get the desired human friendly templated format. Better we could change `appenders` and `layouts` at runtime, without touching the code, from the log4j conf file.

It is WIP because tests still need fixing.

Closes: https://github.com/elastic/elasticsearch/issues/31046

elasticmachine commented:
Pinging @elastic/es-security

albertzaharovits commented on x-pack/plugin/core/src/main/config/log4j2.properties:
```diff
@@ -2,7 +2,34 @@ appender.audit_rolling.type = RollingFile
+                %varsNotEmpty{, "transport.profile":"%enc{%map{transport.profile}}{JSON}"}\
+                %varsNotEmpty{, "rule":"%enc{%map{rule}}{JSON}"}\
+                %varsNotEmpty{, "event.category":"%enc{%map{event.category}}{JSON}"}\
+                }%n
```
This is the format of the `PatternLayout` that is assigned to the rolling file appender of the audit log trail.
It prints each event on one line in the JSON format.
A field with a missing value is not printed (`%varsNotEmpty` function).
`PatternLayout` is preferred to the `JSONLayout` because it allows to define the field order.

albertzaharovits replied:
Good point, will do!

albertzaharovits commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/audit/logfile/CapturingLogger.java:
```diff
@@ -10,24 +10,30 @@
+        return newCapturingLogger(level, null);
+    }
+
+    public static Logger newCapturingLogger(final Level level, StringLayout layout) throws IllegalAccessException {
```
`LoggingAuditTrailTests` now assume that the layout for the `CapturingLogger` is JSON (not strictly a requirement, but it's better to check for events in tests as they would be generated outside the test env), yet `CapturingLogger` is used in other places and we don't want to impose another layout everywhere.

albertzaharovits commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/audit/logfile/CapturingLogger.java:
```diff
@@ -59,28 +65,38 @@ public static boolean isEmpty(final String name) {
+                logLine = new String(getLayout().toByteArray(event), StandardCharsets.UTF_8);
+            } else {
+                logLine = event.getMessage().getFormattedMessage();
+            }
```
Use a layout, if provided, when printing the event. Assumes the layout encodes text as UTF8.

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrail.java:
```diff
@@ -71,63 +73,86 @@
+    public static final String REQUEST_NAME_FIELD_NAME = "request.name";
+    public static final String TRANSPORT_PROFILE_FIELD_NAME = "transport.profile";
+    public static final String RULE_FIELD_NAME = "rule";
+    public static final String OPAQUE_ID_FIELD_NAME = "opaque_id";
```
These are the keys for the `StringMapMessage` and also the field names of the audit event.
If a field name is changed the modification has to be mirrored in the `log4j2.properties` file.
Field names should be [ecs](https://github.com/elastic/ecs) compliant.
@ruflin can you please take a look at them. I know `timestamp` should be `@timestamp` but I wish to change them in one go, so I would appreciate if you can put the stamp of approval over them 😃 
I will pull out a sample of the generated audit trail to make it easy to reason what each field should hold.

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrail.java:
```diff
@@ -71,63 +73,86 @@
+            .boolSetting(setting("audit.logfile.prefix.emit_node_host_address"), false, Property.NodeScope, Property.Dynamic);
+    public static final Setting<Boolean> EMIT_HOST_NAME_SETTING = Setting.boolSetting(setting("audit.logfile.prefix.emit_node_host_name"),
+            false, Property.NodeScope, Property.Dynamic);
+    public static final Setting<Boolean> EMIT_NODE_NAME_SETTING = Setting.boolSetting(setting("audit.logfile.prefix.emit_node_name"), true,
```
node/host name toggles are prefixed with `EMIT_`

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrail.java:
```diff
@@ -71,63 +73,86 @@
-    LocalNodeInfo localNodeInfo;
+    private final ThreadContext threadContext;
+    // fields that all entries have in common
+    EntryCommonFields entryCommonFields;
```
`entryCommonFields` holds the `node` and `host` information that each log entry has.

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrail.java:
```diff
@@ -158,35 +181,35 @@ public LoggingAuditTrail(Settings settings, ClusterService clusterService, Threa
+            restOrigin(request, logEntry);
+            requestBody(request, logEntry);
+            opaqueId(threadContext, logEntry);
+            logger.info(logEntry);
```
enqueues for logging the `StringMapMessage` which has just been built. This is the new recurrent pattern for all event types.

albertzaharovits replied:
you are right, it might be cleaner this way. I will make the change.

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrail.java:
```diff
@@ -493,131 +549,139 @@ public void runAsDenied(Authentication authentication, String action, TransportM
+        if (includeRequestBody) {
+            final String requestContent = restRequestContent(request);
+            if (Strings.hasLength(requestContent)) {
+                logEntry.with(REQUEST_BODY_FIELD_NAME, requestContent);
```
Adds the `request.body` only if enabled and the request content is non-empty.
The request content is escaped in the `log4j2.properties` format file (see [above](https://github.com/elastic/elasticsearch/pull/31931/files#r209964478)), it does not have to be escaped here.

albertzaharovits commented:
run gradle build tests

albertzaharovits commented:
I have no idea why:
https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+pull-request/15135/consoleFull
and
https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+pull-request/15117/consoleFull
failed with
```
16:24:25 FAILURE 0.01s J1 | FileRolesStoreTests.testThatInvalidRoleDefinitions <<< FAILURES!
16:24:25    > Throwable #1: java.lang.AssertionError: 
16:24:25    > Expected: a collection with size <6>
16:24:25    >      but: collection size was <7>
16:24:25    > 	at __randomizedtesting.SeedInfo.seed([A724AFFF473E9AF7:B33BD609698A55D9]:0)
16:24:25    > 	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
16:24:25    > 	at org.elasticsearch.xpack.security.authz.store.FileRolesStoreTests.testThatInvalidRoleDefinitions(FileRolesStoreTests.java:383)
16:24:25    > 	at java.lang.Thread.run(Thread.java:748)
```
and
```
14:14:07   1> [2018-08-14T16:14:07,084][INFO ][o.e.x.s.a.s.FileRolesStoreTests] [testThatRoleNamesDoesNotResolvePermissions] after test
14:14:07 FAILURE 0.00s J0 | FileRolesStoreTests.testThatRoleNamesDoesNotResolvePermissions <<< FAILURES!
14:14:07    > Throwable #1: java.lang.AssertionError: 
14:14:07    > Expected: a collection with size <1>
14:14:07    >      but: collection size was <10>
14:14:07    > 	at __randomizedtesting.SeedInfo.seed([B0FFD9096EC15803:83D76C701FAF4C10]:0)
14:14:07    > 	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
14:14:07    > 	at org.elasticsearch.xpack.security.authz.store.FileRolesStoreTests.testThatRoleNamesDoesNotResolvePermissions(FileRolesStoreTests.java:404)
14:14:07    > 	at java.lang.Thread.run(Thread.java:748)
```
, among others:
```
org.elasticsearch.xpack.security.authz.store.FileRolesStoreTests.testThatInvalidRoleDefinitions
org.elasticsearch.xpack.security.authz.store.FileRolesStoreTests.testParseFileWithFLSAndDLSDisabled
org.elasticsearch.xpack.security.authc.support.DnRoleMapperTests.testParseFileLenient_WhenCannotReadFile
```
so I've pushed be1a6a29a78722f54702cc0fda8fd14ff01ba00b hopping it solves it.

albertzaharovits commented:
test this please

albertzaharovits commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrailTests.java:
```diff
@@ -145,34 +152,70 @@ public void init() throws Exception {
+                "%varsNotEmpty{, \"event.category\":\"%enc{%map{event.category}}{JSON}\"}" +
+                "}%n")
+                .withCharset(StandardCharsets.UTF_8)
+                .build();
```
Duplicate the pattern from `log4j2.properties` here. Not strictly necessary to have the same format but a different one is unwarranted.

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrail.java:
```diff
@@ -493,131 +549,139 @@ public void runAsDenied(Authentication authentication, String action, TransportM
+            final String formattedAddress = NetworkAddress.format(socketAddress.getAddress());
+            logEntry.with(ORIGIN_TYPE_FIELD_NAME, REST_ORIGIN_FIELD_VALUE).with(ORIGIN_ADDRESS_FIELD_NAME, formattedAddress);
         }
+        // fall through to local_node default
```
by default all events originate locally unless one of `restOrigin` or `restOrTransportOrigin` is called to populate the origin details.

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrail.java:
```diff
@@ -493,131 +549,139 @@ public void runAsDenied(Authentication authentication, String action, TransportM
-                new StringBuilder("origin_type=[transport], origin_address=[").append(NetworkAddress.format(address.address().getAddress()))
-                        .append("]")
-                        .toString());
+        // fall through to local_node default
```
by default all events originate locally unless one of restOrigin or restOrTransportOrigin is called to populate the origin details.

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrail.java:
```diff
@@ -813,56 +877,52 @@ public void clusterChanged(ClusterChangedEvent event) {
+            if (newLocalNode != null) {
+                commonFieldsBuilder.put(ORIGIN_ADDRESS_FIELD_NAME, newLocalNode.getHostAddress());
+            }
+            this.commonFields = commonFieldsBuilder.immutableMap();
```
common fields include: `node name`, `host name`, `host ip` and, implicitly originate locally.

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrail.java:
```diff
@@ -493,131 +549,139 @@ public void runAsDenied(Authentication authentication, String action, TransportM
+        }
+        if (request.params().isEmpty() == false) {
+            // TODO this is probably wrong, params are decoded and they might contain the delimiter
+            logEntry.with(URL_QUERY_FIELD_NAME, Strings.collectionToDelimitedString(request.params().entrySet(), "&"));
```
Previously there was a field named `uri` that included the query string. To make it more compliant to [ecs-url-fields]( https://github.com/elastic/ecs#-url-fields) I have split it to `url.path` and `url.query`. What do you think @ruflin ?
However, `url.query` is potentially ambiguous because URL parameters are decoded at this point so they might contain the `&` character.

ruflin replied:
In ECS `url.query` is also just a string and not an array but it's analyzed so Elasticsearch would do the tokenisation for us.

albertzaharovits replied:
okay. I presume it's also decoded, `%26` -> `&`? The trouble is that I am doing concatenation with a character that might also be present in the value (`&`) when decoded.
I am fine with it as it is, I was just raising a potential contestable topic.

ruflin replied:
We don't have any definitions yet on ECS for the encoding. If `&` is a the value, what will your query string look like?

albertzaharovits replied:
> If & is a the value, what will your query string look like?

What I meant here is that the query string parts are decoded such that there are no more % encoded values. Consequently, if a query param or value had `%26` in it, it gets decoded to `&` which is also the separator of individual `param=value` pairs.

Anyway, this is not relevant anymore. I have looked at the ECS schema, and since there is provisioning about encoding - as you said - and since it is expecting the whole query string as text, I suppose the analyzer expects encoded values since otherwise decoded string will be ambiguous - `=` and `&` are special characters they **have** to be encoded.

The outcome is that `url.query` is what follows `?` and until `#`, as is, in the URL. I fathom this is ECS "compliant", isn't it?

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrail.java:
```diff
@@ -196,16 +219,19 @@ public void authenticationSuccess(String realm, User user, String action, Transp
-                    logger.info("{}[transport] [authentication_success]\t{}, {}, realm=[{}], action=[{}], request=[{}]{}",
-                            localNodeInfo.prefix, originAttributes(threadContext, message, localNodeInfo), principal(user), realm, action,
-                            message.getClass().getSimpleName(), opaqueId());
+                    logEntry.with(INDICES_FIELD_NAME, Strings.arrayToCommaDelimitedString(indices.get()));
```
Should `indices` be an array instead of a comma delimited string?

ruflin replied:
What will be the different in the log output? ["a,b,c"] vs ["a", "b", "c"]? If yes, prefer the second one as it does not need additional processing as I would say is as human readable as the first one.

albertzaharovits replied:
> What will be the different in the log output? ["a,b,c"] vs ["a", "b", "c"]?

Close, it is `"indices": "a,b,c"` instead of `"indices": ["a", "b", "c"]`.
There is some convenience of having it as `"a,b,c"`, because this is a string and not an array formatted as a string, but either one is doable! The trickiness lies at the `StringMapMessage` and `PatternLayout` interface.

ruflin replied:
The advantage of `"indices": ["a", "b", "c"]` for us is that if we ingest it into Elasticsearch, it will be a keyword type that contains a list of 3 entries and a query for `b` will return a result. If we have `"indices": "a,b,c"` as keyword a query for `b` will not return a result if we don't split it up on Filebeat side or during ingest time.

albertzaharovits replied:
I have went with the `"indices": ["a", "b", "c"]` alternative. The same for `"user.roles"`.

albertzaharovits commented:
failure such as https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+pull-request/15202/console still elude me. I will log on the jenkins worker after the test fails (and before the mess is wiped out)
jenkins, if you wish, test this please

albertzaharovits commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/audit/logfile/CapturingLogger.java:
```diff
@@ -22,12 +24,13 @@
 public class CapturingLogger {
 
-    public static Logger newCapturingLogger(final Level level) throws IllegalAccessException {
+    public static Logger newCapturingLogger(final Level level, @Nullable StringLayout layout) throws IllegalAccessException {
```
The `CapturingLogger` receives a layout to format the captured messages as strings. Sometimes captured messages are parsed so it would be helpful to be in control of the formatting as well. Otherwise, no layout calls: `event.getMessage().getFormattedMessage()`.

albertzaharovits replied:
I agree 👍

albertzaharovits commented:
Failures such as https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+pull-request/15202/console  were a true programming error.
Logger and appenders have their name registered globally by the log4j library. The names are a combination of class and test method name, making them unique for each test. The class and method names are pulled automatically by inspecting the stacktrace. The error came about when I have inadvertedly burrowed the code inspecting the stacktrace, and instead of returning the test method name it returned another's method name which was not suitable for namespacing loggers. Consequently there were loggers shared between tests, with log events from foreign tests in the "local" test logger.
All is good now!

albertzaharovits commented:
> 11:23:59 Execution failed for task ':x-pack:qa:rolling-upgrade:with-system-key:v6.5.0-SNAPSHOT#oneThirdUpgradedTestCluster#wait'.
11:23:59 > Failed to start elasticsearch: timed out after 30 seconds

run gradle build tests

albertzaharovits commented:
The failure before `(./gradlew :x-pack:qa:ml-native-multi-node-tests:check -Dtests.seed=97BF09639D3216BE`) seems to be reproducible locally. However, it did not on the master branch and then it did not after a local merge. Treating it as a fluke....

albertzaharovits commented:
Previous build failure
```
./gradlew :x-pack:plugin:integTestRunner -Dtests.seed=D27D889127096FD -Dtests.class=org.elasticsearch.xpack.test.rest.XPackRestIT -Dtests.method="test {p0=ml/ml_classic_analyze/Test analyze API with an analyzer that does what we used to do in native code}" -Dtests.security.manager=true -Dtests.locale=es-CO -Dtests.timezone=Australia/Lindeman -Dcompiler.java=10 -Druntime.java=8 -Dtests.rest.blacklist=getting_started/10_monitor_cluster_health/*
```
was due to `null` being present in the array of indices that is passed to `accessGranted` . Not sure if this is a bug or some corner case we're ok with, but throwing NPE in the logging layer is definitely incorrect.
I have addressed it such that `null` values in arrays (`indices` and `user.roles`) are skipped while printing - added tests for this aswell -.

albertzaharovits commented:
`user.roles` and `indices` are now arrays of strings instead of comma separated strings. **The array can be empty as well.** Also, some events don't have this field at all (e.g. `authenticationSuccess`).
While most other fields without values do not get printed at all, IMO ignoring empty lists of indices and roles would be lost information (for the human reader, I don't think it matters to the ingesting cluster).

albertzaharovits commented:
@ruflin I wish we settle on the field names and types so that they are easiest to read by beats and be [ECS](https://github.com/elastic/ecs) compliant.

I went over the schema and have tried to follow it closely. We have agreed we will ship 6.x with **both** log file trail formats enabled, so we are not constrained that the new format be similar to the old one, ie we "maximize" on beats-friendliness alone.
Here is an example of the generated entries:
```
{"@timestamp":"2018-08-28T14:10:44,153", "host.ip":"0.0.0.0:1", "event.type":"transport", "event.action":"tampered_request", "origin.type":"local_node", "origin.address":"0.0.0.0:1", "action":"_action", "request.name":"MockMessage", "opaque_id":"N", "event.category":"elasticsearch-audit"}
{"@timestamp":"2018-08-28T14:10:44,193", "event.type":"transport", "event.action":"authentication_success", "user.name":"_username", "origin.type":"transport", "origin.address":"0.0.0.0", "realm":"GGNBr", "action":"_action", "request.name":"MockMessage", "opaque_id":"OH", "event.category":"elasticsearch-audit"}
{"@timestamp":"2018-08-28T14:10:44,215", "event.type":"ip_filter", "event.action":"connection_denied", "origin.type":"ip_filter", "origin.address":"127.0.0.1", "opaque_id":"fC", "transport.profile":"VrYO", "rule":"deny _all", "event.category":"elasticsearch-audit"}
{"@timestamp":"2018-08-28T14:10:44,238", "host.ip":"0.0.0.0:5", "event.type":"rest", "event.action":"realm_authentication_failed", "user.name":"_principal", "origin.type":"rest", "origin.address":"::1", "realm":"BiIJF", "url.path":"_uri", "url.query":"_param=baz", "event.category":"elasticsearch-audit"}
{"@timestamp":"2018-08-28T14:10:44,257", "host.ip":"0.0.0.0:6", "event.type":"transport", "event.action":"run_as_denied", "user.name":"_username", "user.run_as.name":"running as", "user.realm":"authRealm", "user.run_as.realm":"lookRealm", "user.roles":[], "origin.type":"local_node", "origin.address":"0.0.0.0:6", "action":"_action", "request.name":"MockIndicesRequest", "indices":["xX"], "opaque_id":"X", "event.category":"elasticsearch-audit"}
{"@timestamp":"2018-08-28T14:10:44,483", "host.ip":"0.0.0.0:7", "event.type":"transport", "event.action":"access_granted", "user.name":"_username", "user.realm":"authRealm", "user.roles":["I"], "origin.type":"transport", "origin.address":"0.0.0.0", "action":"_action", "request.name":"MockIndicesRequest", "indices":["zw"], "opaque_id":"Lh", "event.category":"elasticsearch-audit"}
{"@timestamp":"2018-08-28T14:10:44,499", "event.type":"rest", "event.action":"authentication_failed", "origin.type":"rest", "origin.address":"::1", "url.path":"_uri", "url.query":"bar=baz", "request.body":"{ \"key\": \"value\" }", "event.category":"elasticsearch-audit"}
{"@timestamp":"2018-08-28T14:10:44,516", "host.ip":"0.0.0.0:10", "event.type":"transport", "event.action":"authentication_failed", "user.name":"_principal", "origin.type":"rest", "origin.address":"127.0.0.1", "action":"_action", "request.name":"MockIndicesRequest", "indices":["Hx","cpg"], "opaque_id":"UDzw", "event.category":"elasticsearch-audit"}
{"@timestamp":"2018-08-28T14:10:44,552", "host.ip":"0.0.0.0:11", "event.type":"transport", "event.action":"authentication_failed", "origin.type":"local_node", "origin.address":"0.0.0.0:11", "action":"_action", "request.name":"MockMessage", "opaque_id":"paRM", "event.category":"elasticsearch-audit"}
{"@timestamp":"2018-08-28T14:10:44,564", "host.ip":"0.0.0.0:12", "event.type":"rest", "event.action":"authentication_failed", "user.name":"_principal", "origin.type":"rest", "origin.address":"::1", "url.path":"_uri", "url.query":"foo=bar", "event.category":"elasticsearch-audit"}
{"@timestamp":"2018-08-28T14:10:44,575", "host.ip":"0.0.0.0:13", "event.type":"rest", "event.action":"tampered_request", "origin.type":"rest", "origin.address":"127.0.0.1", "url.path":"_uri", "event.category":"elasticsearch-audit"}
{"@timestamp":"2018-08-28T14:10:44,619", "event.type":"transport", "event.action":"anonymous_access_denied", "origin.type":"rest", "origin.address":"127.0.0.1", "action":"_action", "request.name":"MockMessage", "event.category":"elasticsearch-audit"}
{"@timestamp":"2018-08-28T14:10:44,637", "event.type":"transport", "event.action":"run_as_granted", "user.name":"_username", "user.run_as.name":"running as", "user.realm":"authRealm", "user.run_as.realm":"lookRealm", "user.roles":["Gs"], "origin.type":"rest", "origin.address":"127.0.0.1", "action":"_action", "request.name":"MockMessage", "event.category":"elasticsearch-audit"}
{"@timestamp":"2018-08-28T14:10:44,652", "host.ip":"0.0.0.0:19", "event.type":"ip_filter", "event.action":"connection_granted", "origin.type":"ip_filter", "origin.address":"127.0.0.1", "opaque_id":"YVXu", "transport.profile":"VpE", "rule":"allow default:accept_all", "event.category":"elasticsearch-audit"}
{"@timestamp":"2018-08-28T14:10:44,667", "host.ip":"0.0.0.0:20", "event.type":"rest", "event.action":"authentication_success", "user.name":"running as", "user.run_by.name":"_username", "origin.type":"rest", "origin.address":"127.0.0.1", "realm":"UXm", "url.path":"_uri", "url.query":"foo=bar&evac=true", "request.body":"{ \"key\": \"value\" ", "event.category":"elasticsearch-audit"}
{"@timestamp":"2018-08-28T14:10:44,679", "event.type":"transport", "event.action":"access_granted", "user.name":"running_as", "user.run_by.name":"_username", "user.realm":"lookRealm", "user.run_by.realm":"authRealm", "user.roles":["HcL","ncb"], "origin.type":"local_node", "origin.address":"0.0.0.0:21", "action":"internal:_action", "request.name":"MockMessage", "event.category":"elasticsearch-audit"}
{"@timestamp":"2018-08-28T14:10:44,698", "event.type":"transport", "event.action":"tampered_request", "user.name":"_username", "origin.type":"rest", "origin.address":"127.0.0.1", "action":"_action", "request.name":"MockMessage", "event.category":"elasticsearch-audit"}
{"@timestamp":"2018-08-28T14:10:44,724", "event.type":"transport", "event.action":"realm_authentication_failed", "user.name":"_principal", "origin.type":"local_node", "origin.address":"0.0.0.0:23", "realm":"zHj", "action":"_action", "request.name":"MockIndicesRequest", "indices":["mT"], "event.category":"elasticsearch-audit"}
{"@timestamp":"2018-08-28T14:10:44,739", "event.type":"rest", "event.action":"anonymous_access_denied", "origin.type":"rest", "origin.address":"::1", "url.path":"_uri", "request.body":"{ \"key\": \"value\" ", "opaque_id":"J", "event.category":"elasticsearch-audit"}
{"@timestamp":"2018-08-28T14:10:44,755", "host.ip":"0.0.0.0:25", "event.type":"transport", "event.action":"access_granted", "user.name":"_system", "user.realm":"_reserved", "user.roles":["x","DCgJ","am"], "origin.type":"transport", "origin.address":"127.0.0.1", "action":"internal:_action", "request.name":"MockMessage", "opaque_id":"l", "event.category":"elasticsearch-audit"}
{"@timestamp":"2018-08-28T14:10:44,770", "event.type":"transport", "event.action":"access_denied", "user.name":"_username", "user.realm":"authRealm", "user.roles":[], "origin.type":"local_node", "origin.address":"0.0.0.0:26", "action":"_action/bar", "request.name":"MockMessage", "opaque_id":"Ozw", "event.category":"elasticsearch-audit"}
```
can you please point to any shortcomings?
For example: `"host.ip":"0.0.0.0:25"` also contains the port . Another enhancement example might be the inclusion of `"http.request.method"` .

albertzaharovits commented:
@ruflin  Another request that I remembered just now is to have the node name in the rotated log file name.
This is easily achievable, using `${sys:es.logs.node_name}` in the `log4j2.properties` file.
Is this still desirable? If it is, then, should `node.name` still be a field of the log events?

albertzaharovits commented:
test this please

ruflin commented:
Sorry for the late feedback. For the `host.ip`: It should not contain the port as I think in this case it would not be compatible with the ip type in Elasticearch: https://www.elastic.co/guide/en/elasticsearch/reference/current/ip.html I wonder if we should call it `host.address` to not conflict? We should probably have a place for it in ECS.

Overall the fields LGTM. Is there a complete list of fields that can show up available? Would make it easier to review.

For the node name: File name would be nice, but if we can put it into each log line it would make it probably even easier for us and would make sure it's always there. Should we use the `node.id` instead of `node.name` as the name could change?

albertzaharovits commented:
> For the host.ip: It should not contain the port as I think in this case it would not be compatible with the ip type in Elasticearch: https://www.elastic.co/guide/en/elasticsearch/reference/current/ip.html I wonder if we should call it host.address to not conflict? We should probably have a place for it in ECS.

I am leaning towards removing the port and keeping the `host.ip` field and not adding `host.address` in ECS. `ip` is an attribute of the host but the `port` is not. `port` could be an attribute of the `service` or maybe of the `url`. But, in any case, I prefer to keep `port` out of the way altogether because it is the bound port, and not the advertised one and this could create confusion - arguably `host.ip` is also the bound and not the advertised ip, but it is truthfully an attribute of the host. Alternatively, I could have the port as another attribute `host.port`, or as `url.port`, but again, maybe there are more reliable ways to pull this from.
Let me know what you think.

> For the node name: File name would be nice, but if we can put it into each log line it would make it probably even easier for us and would make sure it's always there. Should we use the node.id instead of node.name as the name could change?

Okay. I will add the `node.id` attribute **alongside** the `node.name` for each log line. `node.id` is enabled by default and `node.name` is disabled by default. I will adjust the file name (for the new audit trail) to also contain the node name.
Let me know if that's alright with you.

> Is there a complete list of fields that can show up available?

Ideally each field for each event type should be documented, but I will address this later. I will followup with a list with all field names, their types, and a typical value, but note that some fields might be present for certain event types and missing for others.
Stay posted!

albertzaharovits commented:
@ruflin here are the JSON fields in plain splendor. Let me know if it's still LGTM 😃 

| JSON Field name | Example values | Details |
-------------------|-----------------|--------|
| "origin.type" | "rest" or "transport" or "ip_filter" | It is "rest" if the event is "in connection to" a REST  request, i.e. is part of the handling of a rest request. |
| "origin.address" | "81.128.11.2:60001" | The remote address and port of the first hop. If this is a REST request this is the proxy, otherwise the coordinating node, etc.). |
| "node.name" | "node01-prod" | Node name from the settings (https://www.elastic.co/guide/en/elasticsearch/reference/current/node.name.html). |
| "host.ip" | "0.0.0.0" | The local bound ip (no port, right?). |
| "host.name" | "localhost" | Just the hostname. |
| "event.category" | "elasticsearch-audit" | Fixed value. |
| "event.type" | "rest" or "transport" | This has more internal meaning. A REST request is translated to transport messages. This field indicates which "layer" has generated this log line. |
| "event.action" | "authentication_failed" or "access_granted" or "run_as_granted" etc | The "action" that is really being logged. |
| "user.name" | "azaha" | The subject as authenticated by a realm. |
| "user.run_by.name" | "kibana" | The original authenticated user who is impersonating the "user.name" (run as). |
| "user.run_as.name" | "user" | If this event is from a run as, it is the user to be impersonated as. |
| "user.realm" | "saml1" | The name of the realm that authenticated "user.name". |
| "user.run_by.realm" | "ldap2" | The realm that authenticated the impersonating user. |
| "user.run_as.realm" | "native" | If this is event is for a run as, this is the realm the impersonated user is looked up from. |
| "user.roles" | ["role1", "admin"] | The array of roles of the user. This are the roles that are granting privileges. |
| "realm" | "pki2" | The name of the realm that had generated "authentication_failed" or "authentication_success" event. (The user is not yet authenticated). |
| "url.path" | "index/type/id" | the URI component between port and query string ("?"). This is percent (URL) encoded. |
| "url.query" | "refresh=true&foo=bar" | the URI componenet after the "?" and before the fragment ("#"). |
| "request.body" | "\{ \"doc\": \"text\" \}" | JSON escaped request body content. |
| "action" | "indices:data/write/index" or "cluster:monitor/state" etc | This is steeped in internal meaning. An action is the most granular operation which is authz and this identifies it in a namespaced way. |
| "indices" | ["idx1", "idx2"] | The list of indices that the "action" is acting upon. |
| "request.name" | "ClusterStatsNodeRequest" | If the event is about a transport message this is the name of the request class (internal significance, similar to how you identify REST requests by url path). |
| "transport.profile" | "profile1" | the name of the profile if this is a "connection_granted" or "connection_denied" event. |
| "rule" | "deny all" | if this is an event from the ip filter layer, this represents the name of the applied rule. |
| "opaque_id" | "xHd8" | Random string the user submitted during as thee request header "X-Opaque-Id". |

jaymode commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrail.java:
```diff
@@ -416,53 +459,70 @@ public void tamperedRequest(String action, TransportMessage message) {
-                    logger.info("{}[transport] [tampered_request]\t{}, {}, action=[{}], request=[{}]{}", localNodeInfo.prefix,
-                            originAttributes(threadContext, request, localNodeInfo), principal(user), action,
-                            request.getClass().getSimpleName(), opaqueId());
+                    logEntry.with(INDICES_FIELD_NAME, toQuotedJsonArray(indices.get()));
```
There is no way to have the log4j2 appender do the formatting for us?

albertzaharovits replied:
Unfortunately no ☹️ 
[Log4j2 layout pattern functions](https://logging.apache.org/log4j/2.x/manual/layouts.html#PatternLayout) are fairly low level, comparable to printf ones. What we need is "format this as a JSON array field value" (escape then quote every value and concatenate with ,), but what we have instead is just the JSON escape function.
I disliked this too, not being able to have a single point were we handle the formatting, but I dislike more the comma separated strings because we are departing from JSON, the motive for all of this.
Also, array values are the only exception, we are not going to add others constructs that need formatting in the java code.

jaymode commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrail.java:
```diff
@@ -813,56 +914,52 @@ public void clusterChanged(ClusterChangedEvent event) {
-                final String address = localNode != null ? localNode.getHostAddress() : null;
-                if (address != null) {
-                    builder.append("[").append(address).append("] ");
+            final MapBuilder<String, String> commonFieldsBuilder = new MapBuilder<>();
```
just use `new HashMap<>()` and wrap with `Collections.unmodifiableMap()`

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrail.java:
```diff
@@ -71,63 +75,86 @@
 
 public class LoggingAuditTrail extends AbstractComponent implements AuditTrail, ClusterStateListener {
 
+    // WARNING: changing any of this requires changing the log4j2.properties file too
```
> What is the behavior if I use the wrong name in log4j2.properties and try to add log message, does it fail for the field not found or it is ignored?

All key references to the `StringMapMessage` dictionary are wrapped in the log4j2 function `%varsNotEmpty{}`. Because of this, missing fields will not be printed. So, if you change the name here but not in the properties file (or vice-versa), the field will be silently ignored.

> I am not sure what the purpose is for StringMapMessage#validate(key, value) (by default it does not have any validation) something that could be used to enforce validation on field, I do not feel strongly about this as this adds overhead to the logging but still raising it so the we do not miss it.

That's an interesting idea, but I don't think `StringMapMessage#validate(key, value)` can help us achieve this validation. That's because `StringMapMessage` building is oblivious of the layout.
What we could try instead, is to get a hold of the layout at runtime, get the pattern string and check if each map references in the pattern layout is backed by a field we add to the `StringMapMessage`.
Might turn out to be too complicated, but I will try and report back.

albertzaharovits replied:
@bizybot I've been mulling over your suggestion to have some sort of validation of the pairings between field names in the java code and field names in the properties file. I have pushed https://github.com/elastic/elasticsearch/pull/31931/commits/2a3857db5f9a54c53817c077df155846838e1b0f specifically in relation to this. What follows is my reasoning:

First off, the properties file is there to allow the administrator to configure the format. For example it is perfectly reasonable to expect that some fields will be disabled by the administrator by changing the format. However, it is true that a typo might slip in and some field will not be printed, as otherwise expected. In this case, maybe an ERROR log entry (in the usual log) would be a nice to have; or even refusing to apply the config change. However, this is not that simple in practice, mainly because there is no such feedback loop from the log4j library itself. Also, it is too complex to parse the properties file ourselves because it might contain what's called a [Pattern Selector](https://logging.apache.org/log4j/2.x/manual/layouts.html#PatternLayout). This is basically a more versatile pattern, complete with scripting to select the pattern and what not...

Therefore I have abandoned the holy grail of trying to protected from administrator errors and aimed to solve the simpler problem of protecting against our own mistakes. In this respect, I have also avoided parsing the pattern from the config and pairing each variable reference to a field in the code (this would have been another test method). The reason for this is that it turned out to be messy and brittle: there are quite some functions available and the pattern can get complex. In addition it required to have a list of all the fields on the java code; just messy...

What I have decided to go with is to have the `LoggingAuditTrailTests` pull the layout that they're using from the actual `log4j2.properties` file. Now, if we ship with an error in the config some tests will catch it. Moreover, this change avoids the pattern duplication that was used previously (the tests programmatically configure a Mockup appender with the legit layout). Furthermore, having tests run over the true formatting, acknowledges the fact that formatting is split between the code and log4j (https://github.com/elastic/elasticsearch/pull/31931#discussion_r215213222) and we're testing the final composed format of the log entries.

Hope this assuages your worries about config file discordance.
Let me know what you think!

albertzaharovits commented:
@bizybot @jaymode Thank you for the suggestions! 👍  I have addressed them all.
May you please take another look?

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrail.java:
```diff
@@ -158,35 +185,36 @@ public LoggingAuditTrail(Settings settings, ClusterService clusterService, Threa
+        if (events.contains(AUTHENTICATION_SUCCESS) && eventFilterPolicyRegistry.ignorePredicate()
+                .test(new AuditEventMetaInfo(Optional.of(user), Optional.of(realm), Optional.empty(), Optional.empty())) == false) {
+            final StringMapMessage logEntry = new LogEntryBuilder()
+                    .with(EVENT_TYPE_FIELD_NAME, REST_ORIGIN_FIELD_VALUE)
```
I don't like it. I feel we're overdoing it as it provides no benefits.
Ideally I would've used a dictionary instead of builder methods, as it reduces the code, but as you pointed out is much neater to use builder methods instead of static functions with a map as parameter (OOP FTW).

But if you think the MACRO name is not illustrative we could change them?

ruflin commented:
* Port: +1 on removing port for now. Adding a field later is easy, removing it is kind of a breaking change.
* Node.id: Having it by default SGTM as I assume that is what we will use for correlation.
* Documented fields: This is really nice. Is this manually generated? If possible this should end up in the docs. It's one of the big issues with most logs that there are no docs on what will be in the docs itself.
* origin: Not in ECS yet but I like the usage and I don't see a conflict coming here.
* realm: Should not conflict as long as we don't have a realm object.
* action: As this is very specific to Elasticsearch, I would put it under `elasticsearch.action`. But it's something we could also do during ingest time.
* indices: What about `elasticsearch.index.name`: No need for plural as in ES it's always an array. Same as for action, we can also handle it during ingest time.
* Same as for indices adn action also applies to rule and opaque_id but we can / should do that at ingest time.

@albertzaharovits Do you seen an issue that Filebeat would still do some potential transformations on the event during ingest time? We do similar things when we fetch metrics and prefix it with the service to prevent conflicts. It would not make sense in an Elasticsearch API to prefix everything and I think the same is true for a log.

albertzaharovits commented:
Alrighty, based on the feedback from @ruflin I have:
* removed the port in the "host.ip" field
* kept the port on the "origin.address" field
* added "node.id" which defaults to printed (`audit.logfile.prefix.emit_node_id`: true)
* "node.name" is no longer printed implicitly (`audit.logfile.prefix.emit_node_name`: false)
* removed "event.category": "elasticsearch-audit" because this is a constant field
that should be added during ingest

> Documented fields: This is really nice. Is this manually generated? If possible this should end up in the docs. It's one of the big issues with most logs that there are no docs on what will be in the docs itself.

Yep, they're handcrafted. They will definitely be in the documentation (there is already [documentation](https://www.elastic.co/guide/en/elastic-stack-overview/6.4/audit-event-types.html) for it, but it is stale).

> Do you seen an issue that Filebeat would still do some potential transformations on the event during ingest time? We do similar things when we fetch metrics and prefix it with the service to prevent conflicts. It would not make sense in an Elasticsearch API to prefix everything and I think the same is true for a log.

No, I don't take any issue with this.

> indices: What about elasticsearch.index.name: No need for plural as in ES it's always an array. Same as for action, we can also handle it during ingest time.
Same as for indices adn action also applies to rule and opaque_id but we can / should do that at ingest time.

I think all top level fields which are not part of ECS have to/should be prepended during ingest with the appropriate namespace.

jaymode commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrailFilterTests.java:
```diff
@@ -1554,7 +1566,7 @@ public void testIndicesFilter() throws Exception {
 
     private <T> List<T> randomListFromLengthBetween(List<T> l, int min, int max) {
-        assert (min >= 0) && (min <= max) && (max <= l.size());
+        assert min >= 0 && min <= max && max <= l.size();
```
keep the parentheses?

albertzaharovits replied:
I though I can spot when Eclipse does these check while validating the commit diff, but it looks like I can't. I have removed the check. I have long entertained the idea of using vim and things like this are only stoking it...

jaymode commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrail.java:
```diff
@@ -320,58 +353,63 @@ public void authenticationFailed(String realm, AuthenticationToken token, String
         final User user = authentication.getUser();
         final boolean isSystem = SystemUser.is(user) || XPackUser.is(user);
-        if ((isSystem && events.contains(SYSTEM_ACCESS_GRANTED)) || ((isSystem == false) && events.contains(ACCESS_GRANTED))) {
+        if (isSystem && events.contains(SYSTEM_ACCESS_GRANTED) || isSystem == false && events.contains(ACCESS_GRANTED)) {
```
I did not see these get added back?

albertzaharovits commented:
@ruflin I am going to merge this. There are no outstanding moot points, I have addressed them all.
If new ideas sprout on how to enhance the new log format I think they suit in new issues.
